### ISSUE TITLE
[1.3.1 - Small Incompatibility] GeneralImprovements UseBetterMonitors option

### ISSUE THREAD

**Issue by SeedyMcSeedface** [lammas321/lammOS-LethalCompany#5]
I'm mostly putting this here as a question on if this is something that would be able to be patched to be compatible.

The helmet cam view (at least from [OpenBodyCams](https://thunderstore.io/c/lethal-company/p/Zaggy1024/OpenBodyCams/) - have not tested any of the others) does not work when using the monitor/v command and instead it only shows the ship exterior camera when the UseBetterMonitors option from [GeneralImprovements](https://thunderstore.io/c/lethal-company/p/ShaosilGaming/GeneralImprovements/) is enabled.

This is with the UseBetterMonitors option set to false. Displays the helmet cam view as expected.
![1966720_20240207214622_1](https://github.com/lammas321/lammOS-LethalCompany/assets/155770932/6abfd667-7a03-462c-9e03-2412c3abc916)
LogOutput: [UseBetterMonitors_False_LogOutput.log](https://github.com/lammas321/lammOS-LethalCompany/files/14203757/UseBetterMonitors_False_LogOutput.log)

With UseBetterMonitors option set to true. Displays the ship exterior cam instead.
![1966720_20240207214957_1](https://github.com/lammas321/lammOS-LethalCompany/assets/155770932/2d969e9d-359c-4de7-be18-bca4000b8651)
![1966720_20240207215002_1](https://github.com/lammas321/lammOS-LethalCompany/assets/155770932/e7874fc8-696d-4c0d-96ac-dee25d7dff30)
LogOutput: [UseBetterMonitors_True_LogOutput.log](https://github.com/lammas321/lammOS-LethalCompany/files/14203764/UseBetterMonitors_True_LogOutput.log)

Code for profile used in testing: 018d86ac-05eb-2c5d-dbdd-8fe2a8f29a33

I don't mind having to have the additional monitors from GeneralImprovements disabled - but if it's possible to be patched that would be appreciated.

If there's anything else I can do or provide please let me know and I will do my best to assist.

**Comment by lammas321** [comment]
This incompatibility with OpenBodyCameras' camera not showing up on the terminal will be fixed in v1.4.0, which should be coming out relatively soon. As for other helmet/body camera mods, such as the one's by RickArg and CapyCat my mod supports, they don't support the better monitors from GeneralImprovements (they don't have any compatibility programmed in like OpenBodyCameras does, and thus don't render on the bottom right monitor at all) so they also won't render on the terminal.

Once v1.4.0 comes out, OpenBodyCameras will be the only one of the three to work with lammOS while GeneralImprovements' UseBetterMonitors config option is enabled (the other two will result in lammOS logging a warning to the console). OpenBodyCameras' camera will also only work if it's on the bottom right monitor. I could make it work on any of the monitors, but that would require some more work and looking into the issue, and I assume that's where most people will have the camera anyway. OpenBodyCameras' monitor index can be changed by modifying its GeneralImprovementsBetterMonitorIndex config option. If it's not set to 0 or 14, it won't be visible on the terminal and lammOS will log a warning.

**Comment by SeedyMcSeedface** [comment]
Great to hear! Looking forward to it.

Also another small thing I noticed that also seemed to have prevented the OpenBodyCamera view from working is in [CozyImprovements](https://thunderstore.io/c/lethal-company/p/Spyci/CozyImprovements/)
Having the option to keep the terminal monitor on also breaks it and cause the external ship camera to display.

That one's even more minor and less of an issue, but it did surprise me when i found out it was causing the issue as well when I was going through testing stuff.

**Comment by lammas321** [comment]
From looking a glance at CozyImprovements' code, I believe this update should also fix that issue. That mod just makes sure the terminal's screen doesn't turn off, so it was showing exactly what it was designed to do, even though it wasn't correct. Shall I close this issue now that v1.4.0 is out?

**Comment by SeedyMcSeedface** [comment]
I'll check things in a bit when I can and confirm, unless you've confirmed yourself already.

**Comment by SeedyMcSeedface** [comment]
So now when the terminal always on option from CozyImprovements is enabled, it now breaks the ability to use the terminal entirely.
This was tested with UseBetterMonitors enabled.
I'll check without it being enabled it a moment.

[LogOutput.log](https://github.com/lammas321/lammOS-LethalCompany/files/14216335/LogOutput.log)
018d8b81-6417-f22e-4d7b-a901d4bcd089

**Comment by SeedyMcSeedface** [comment]
Okay yeah same thing with the UseBetterMonitors option disabled

Other than that, OpenBodyCams works with UseBetterMonitors enabled.

**Comment by lammas321** [comment]
I just looked at the log you sent and the code for both CozyImprovements and my mod and immediately understood the issue haha. I'll make a quick fix update. Basically, CozyImprovements sets the text on the terminal the moment the terminal is created (after the Start method is called), just before my mod recognizes that the terminal even exists (which would happen just after CozyImprovement's Start method patch). This with my mod detecting CozyImprovements' change to the terminal's text leads to the null reference exception, as my mod still believes the terminal is null.

**Comment by SeedyMcSeedface** [comment]
I found something else causing an issue in my group's mod pack, but I'm not quite sure what it is yet.
I'm trying to figure it out, and will send a log in a bit.

**Comment by lammas321** [comment]
I'll pause on that update for now then, let me know when you figure it out

**Comment by SeedyMcSeedface** [comment]
Here's the log [LogOutput.log](https://github.com/lammas321/lammOS-LethalCompany/files/14216513/LogOutput.log)

I feel like I might have an idea on what's causing it, going to test that by disabling it now.

**Comment by lammas321** [comment]
I feel like it has to do with this part of the log, which mod adds the 'Shrimp' entity?
```
[Warning:    lammOS] The bestiary entry 'Shrimp' could not successfully be matched to an entity. Adding as a standalone bestiary entry.
[Error  : Unity Log] ArgumentException: An item with the same key has already been added. Key: Shrimp
Stack trace:
System.Collections.Generic.Dictionary`2[TKey,TValue].TryInsert (TKey key, TValue value, System.Collections.Generic.InsertionBehavior behavior) (at <787acc3c9a4c471ba7d971300105af24>:IL_015A)
System.Collections.Generic.Dictionary`2[TKey,TValue].Add (TKey key, TValue value) (at <787acc3c9a4c471ba7d971300105af24>:IL_0000)
lammOS.Variables.Variables.PostLoadingEntities () (at <a0001703a0b149f4a2b0fe5b2ecbef91>:IL_00A3)
lammOS.lammOS.Load () (at <a0001703a0b149f4a2b0fe5b2ecbef91>:IL_0019)
lammOS.lammOS.Setup () (at <a0001703a0b149f4a2b0fe5b2ecbef91>:IL_0000)
lammOS.Patches.Patches+TerminalPatches.PostStart (Terminal& __instance) (at <a0001703a0b149f4a2b0fe5b2ecbef91>:IL_0007)

**Comment by SeedyMcSeedface** [comment]
It's for sure something that relies on [LethalLevelLoader](https://thunderstore.io/c/lethal-company/p/IAmBatby/LethalLevelLoader/).
When I disabled that and everything that relies on it, it worked like expected.
The Shrimp entity comes from [LC Office](https://thunderstore.io/c/lethal-company/p/Piggy/LC_Office/)

**Comment by SeedyMcSeedface** [comment]
It's not LethalLevelLoader itself, going to try enabling LC Office now.

**Comment by SeedyMcSeedface** [comment]
Okay yeah it is LC Office, I'll go through and check my other mods that rely on LethalLevelLoader to be sure though.

[LogOutput.log](https://github.com/lammas321/lammOS-LethalCompany/files/14216571/LogOutput.log)

**Comment by SeedyMcSeedface** [comment]
Yeah, it looks like it's LC Office, as it's the only thing relying on LLL that caused the issue when enabled.

**Comment by lammas321** [comment]
The error occurs because LC_Office calls Enemies.RegisterEnemy (an api method within LethalLib to add enemies easier) with the terminal bestiary entry for the Shrimp entity, adding it to the list of entries. But then (probably either code they forgot to remove or didn't know was unnecessary) they add the bestiary entry again themself, causing there to be 2 entries that are exactly identical, leading to the error in lammOS

**Comment by SeedyMcSeedface** [comment]
Is that something the maker of LC Office would need to fix then? Or is there something you're able to do?

**Comment by lammas321** [comment]
I can prevent the error by checking if there's multiple of the same entry (I never thought I'd need to check so I didn't have a check there in the first place, but I'll be adding it now), but to prevent unintended behavior it'd be best for them to fix that themself anyway.

**Comment by SeedyMcSeedface** [comment]
I have made a report on the GitHub for LC Office and have made a direct link to your reply on why the error occurs.

**Comment by SeedyMcSeedface** [comment]
I have no idea why, but I've now started having issues with AdvancedCompany and lammOS. I'm not sure which is the main culprit, but I still experienced this on a fresh profile with only lammOS, AdvancedCompany, and their requirements enabled.

When I load into the game I see this come up in the console:

```
[Info   :AdvancedCompany] Searching for all moons...
[Error  : Unity Log] InvalidOperationException: Sequence contains no matching element
Stack trace:
System.Linq.Enumerable.First[TSource] (System.Collections.Generic.IEnumerable`1[T] source, System.Func`2[T,TResult] predicate) (at <dab7f68612224ba3ae40f651d44f9d4c>:0)
AdvancedCompany.Game.Manager+Moons.SearchMoons () (at G:/Unity/AdvancedCompany/UpgradeCompany/Game/Manager/Moons.cs:361)
AdvancedCompany.Game.Manager.Update (Terminal __instance) (at G:/Unity/AdvancedCompany/UpgradeCompany/Game/Manager.cs:196)
(wrapper dynamic-method) Terminal.DMD<Terminal::Update>(Terminal)
```

When I go to enter the store it'll display this message on the terminal
![1966720_20240209083454_1](https://github.com/lammas321/lammOS-LethalCompany/assets/155770932/de501cc2-b41d-4466-a643-41147dcd9401)

When that message displays, I see the following in the console:
```
[Error  :    lammOS] An error occurred executing the command: 'store'
System.InvalidOperationException: Failed to compare two elements in the array. ---> System.Collections.Generic.KeyNotFoundException: The given key 'UnlockableItem' was not present in the dictionary.
  at System.Collections.Generic.Dictionary`2[TKey,TValue].get_Item (TKey key) [0x0001e] in <787acc3c9a4c471ba7d971300105af24>:0 
  at AdvancedCompany.Game.Manager+Items.GetUnlockablePrice (UnlockableItem unlockable) [0x00017] in G:\Unity\AdvancedCompany\UpgradeCompany\Game\Manager\Items.cs:300 
  at AdvancedCompany.Game.Manager.GetNodeCost (TerminalNode node) [0x00138] in G:\Unity\AdvancedCompany\UpgradeCompany\Game\Manager.cs:429 
  at (wrapper dynamic-method) lammOS.Variables.Variables.DMD<lammOS.Variables.Variables::GetUnlockableCost>(string)
  at lammOS.Commands.Commands+StoreCommand+<>c.<GenerateUpgrades>b__2_0 (lammOS.Variables.Variables+PurchasableUnlockable x, lammOS.Variables.Variables+PurchasableUnlockable y) [0x00025] in <a0001703a0b149f4a2b0fe5b2ecbef91>:0 
  at System.Collections.Generic.ArraySortHelper`1[T].InsertionSort (T[] keys, System.Int32 lo, System.Int32 hi, System.Comparison`1[T] comparer) [0x0002a] in <787acc3c9a4c471ba7d971300105af24>:0 
  at System.Collections.Generic.ArraySortHelper`1[T].IntroSort (T[] keys, System.Int32 lo, System.Int32 hi, System.Int32 depthLimit, System.Comparison`1[T] comparer) [0x0004b] in <787acc3c9a4c471ba7d971300105af24>:0 
  at System.Collections.Generic.ArraySortHelper`1[T].IntrospectiveSort (T[] keys, System.Int32 left, System.Int32 length, System.Comparison`1[T] comparer) [0x00013] in <787acc3c9a4c471ba7d971300105af24>:0 
  at System.Collections.Generic.ArraySortHelper`1[T].Sort (T[] keys, System.Int32 index, System.Int32 length, System.Comparison`1[T] comparer) [0x00000] in <787acc3c9a4c471ba7d971300105af24>:0 
   --- End of inner exception stack trace ---
  at System.Collections.Generic.ArraySortHelper`1[T].Sort (T[] keys, System.Int32 index, System.Int32 length, System.Comparison`1[T] comparer) [0x00023] in <787acc3c9a4c471ba7d971300105af24>:0 
  at System.Collections.Generic.List`1[T].Sort (System.Comparison`1[T] comparison) [0x00013] in <787acc3c9a4c471ba7d971300105af24>:0 
  at lammOS.Commands.Commands+StoreCommand.GenerateUpgrades () [0x000ae] in <a0001703a0b149f4a2b0fe5b2ecbef91>:0 
  at lammOS.Commands.Commands+StoreCommand.Execute (Terminal terminal, System.String input) [0x00005] in <a0001703a0b149f4a2b0fe5b2ecbef91>:0 
  at lammOS.NewTerminal.NewTerminal.HandleCommand (System.String input) [0x000f4] in <a0001703a0b149f4a2b0fe5b2ecbef91>:0
```

The weirdest part is that this wasn't happening about 6-ish hours ago and nothing in my group's mod list has updated since then.

As usual, profile code and LogOutput below

018d8e17-8d08-06d7-3add-59340abaf996
[LogOutput.log](https://github.com/lammas321/lammOS-LethalCompany/files/14222343/LogOutput.log)

Edit: Formatting

**Comment by lammas321** [comment]
This error was caused by a slight error on my part, I'll be including the fix in the next update. I removed a bunch of the default terminal keywords, as most of their functionality is available through lammOS commands, but at some point in time after I remove them, AdvanedCompany tries to use some of the ones I removed

**Comment by lammas321** [comment]
Just released v1.4.1 fixing all of these issues that you've mentioned. If you find any others, you can either mention them here or let me know to close this issue so you can open another that'd be more on topic. You can also always contact me through this [lammOS Discord thread](https://discord.com/channels/1168655651455639582/1196941743673847938) in the [Lethal Company Modding Discord server](https://discord.com/invite/lcmod), I'm able to see and respond to reports there sooner than I would here

**Comment by SeedyMcSeedface** [comment]
I think this is good to close. If I find or notice anything else I'll put up another report.
Thanks for the work you do - absolutely love it. Cheers!
